keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,863 @@
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New Resource
+
+```
+type ClusterCIDRConfig struct {
+    # An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    # +required
+    CIDRBlock     string
+
+    # This defines which nodes the config is applicable to. An empty selector
+    # can be applied to any node.
+    # +optional
+    NodeSelector    *v1.LabelSelector
+
+    # Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    # Users would have to ensure that the kubelet doesn't try to schedule
+    # more pods than are supported by the node's netmask (i.e. the kubelet's
+    # --max-pods flag)
+    # +required
+    PerNodeMaskSize int
+}
+```
+
+#### Expected Behavior
+
+-   Each node will be assigned up to one range from each `FamilyType`.
+    -   **TODO:** what is the tie-break if multiple ranges match
+-   An empty `NodeSelector` functions as a default that applies to all nodes.
+    This should be the fall-back and not take precedence if any other range
+    matches.
+-   `CIDRBlock`, and `PerNodeMaskSize` are immutable after creation.
+-   Reject any changes that would change existing node allocations:
+    -   Can not delete a `CIDRRangeConfig` that is in use by a node
+    -   Can not modify a `NodeSelector` to deselect a node that is using that
+        range.
+
+#### Example Config
+
+```
+[
+{
+    # For existing clusters this is the same as ClusterCIDR
+    CIDRBlock:     "10.0.0.0/8",
+    # Default for nodes not matching any other rule
+    NodeSelector:    nil,
+    # For existing API this is the same as NodeCIDRMaskSize
+    PerNodeMaskSize: 24,
+},
+{
+    CIDRBlock:     "172.16.0.0/14",
+    # Another range, also allocate-able to any node
+    NodeSelector:    nil,
+    PerNodeMaskSize: 24,
+},
+{
+    CIDRBlock:     "10.0.0.0/8",
+    NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+    PerNodeMaskSize: 26,
+},
+{
+    CIDRBlock:     "192.168.0.0/16",
+    NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+    PerNodeMaskSize: 26,
+},
+{
+    CIDRBlock:     "5.2.0.0/16",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 20,
+},
+{
+    CIDRBlock:     "fd12:3456:789a:1::/64",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 112,
+},
+...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{“np”: “np1”} --> "10.0.0.0/26"
+{“np”: “np2”} --> "192.16.0.0/26"
+{“np”: “np3”} --> "5.2.0.0/20", "fd12:3456:789a:1::/112"
+{“np”: “np4”} --> "172.16.0.0/24"
+```
+
+### Controller
+
+Implement a new
+[NodeIPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam)
+The controller will set up watchers on the `ClusterCIDRConfig` objects and the
+`Node` objects.
+
+#### Data Structures
+
+We want to use a set of radix trees to store allocation. This is a commonly used
```

Comment:
I wonder if we even need to go this far.  We're talking about O(thousands) of Nodes - 15000 is the largest I know of, but even assuming 64K, we can probably afford to optimize for maintenance.   Even a simple sorted array might cut it.  I don't know what libs exist for this in Go, but let's not shoot the moon.

Reply:
maps will work, you have an example and some benchmarks here https://github.com/kubernetes/kubernetes/pull/90184

Reply:
That sounds good. I reworked the section and added a link to your benchmarks.